File: que46.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backTrack(result,new ArrayList<>(),nums);
        return result;
    }
    private void backTrack(List<List<Integer>> result, ArrayList<Integer> tempList,int[] nums){
        if(tempList.size() == nums.length){
            result.add(new ArrayList<>(tempList));
            return;
        }
        for(int number : nums){
            if(tempList.contains(number)){
                continue;
            }
            tempList.add(number);
            backTrack(result,tempList,nums);
            tempList.remove(tempList.size() - 1);
        }
    }
}

File: que46.cpp
class Solution {
public:
    void getAllPerm(vector<int>& nums,int idx,vector<vector<int>> &allPerm){
        if(idx==nums.size()){
            allPerm.push_back({nums});
            return;
        }
        for(int i=idx;i<nums.size();i++){
            swap(nums[i],nums[idx]);
            getAllPerm(nums,idx+1,allPerm);
            swap(nums[i],nums[idx]);
        }

    }
    vector<vector<int>> permute(vector<int>& nums) {
       vector<vector<int>> allPerm;
       getAllPerm(nums,0,allPerm);
       return allPerm; 
    }
};
